crates/ar-cli/src/db.rs

```diff
@@ -0,0 +1,151 @@
+        let tx_hash_hex = hex::encode(&self.tx_hash);
+        let mac_salt_hex = hex::encode(&self.mac_salt);
+        let mac_commitment_hex = hex::encode(&self.mac_commitment);
+        let viewing_key_hex = self
+            .viewing_key
+            .as_ref()
+            .map_or("None".to_string(), hex::encode);
+        write!(
+            f,
+            "Event {{\n  tx_hash: 0x{},\n  block_number: {},\n  mac_salt: 0x{},\n  mac_commitment: 0x{},\n  viewing_key: 0x{}\n}}",
+            tx_hash_hex, self.block_number, mac_salt_hex, mac_commitment_hex, viewing_key_hex
+        )
+    }
+}
+
+pub fn create_events_table(connection: &Connection) -> Result<(), rusqlite::Error> {
+    connection.execute(
+        "CREATE TABLE IF NOT EXISTS events (
+            tx_hash BLOB PRIMARY KEY,
+            block_number INTEGER NOT NULL,
+            mac_salt BLOB NOT NULL,
+            mac_commitment BLOB NOT NULL,
+            viewing_key BLOB
+        )",
+        (),
+    )?;
+    Ok(())
+}
+
+pub fn upsert_event(connection: &Connection, event: Event) -> Result<(), rusqlite::Error> {
+    connection.execute(
+        "REPLACE INTO events (tx_hash, block_number, mac_salt, mac_commitment, viewing_key) VALUES (?1, ?2, ?3, ?4, ?5)",
+        (&event.tx_hash, &event.block_number, &event.mac_salt, event.mac_commitment, event.viewing_key),
+    )?;
+
+    Ok(())
+}
+
+/// Query event by it's tx_hash
+pub fn query_event(connection: &Connection, tx_hash: &[u8; 32]) -> Result<Event, rusqlite::Error> {
+    connection.query_row(
+        "SELECT tx_hash, block_number, mac_salt, mac_commitment, viewing_key FROM events WHERE tx_hash = ?1",
+        [tx_hash],
+        |row|  {
+            Ok(Event {
+                tx_hash: row.get(0)?,
+                block_number: row.get(1)?,
+                mac_salt: row.get(2)?,
+                mac_commitment: row.get(3)?,
+                viewing_key: row.get(4)?,
+            })
+        })
+}
+
+/// Query events by optional viewing_key
+pub fn query_events(
+    connection: &Connection,
+    viewing_key: Option<Vec<u8>>,
+) -> Result<Vec<Event>, rusqlite::Error> {
+    let statement = match viewing_key {
+        Some(_) => "SELECT tx_hash, block_number, mac_salt, mac_commitment, viewing_key FROM events where viewing_key = ?1", 
+        None => "SELECT tx_hash, block_number, mac_salt, mac_commitment, viewing_key FROM events", 
+    };
+
+    let mut query = connection.prepare(statement)?;
+
+    let f = |row: &Row| -> Result<Event, rusqlite::Error> {
+        Ok(Event {
+            tx_hash: row.get(0)?,
+            block_number: row.get(1)?,
+            mac_salt: row.get(2)?,
+            mac_commitment: row.get(3)?,
+            viewing_key: row.get(4)?,
+        })
+    };
+
+    let results = match viewing_key {
+        Some(key) => query.query_map([key], f)?,
+        None => query.query_map([], f)?,
+    };
+
+    let mut events = vec![];
+    for r in results {
+        events.push(r?);
+    }
```

Comment:
afaik, vec of  result can be collected to result of vector

Reply:
good spot!